### ISSUE TITLE
streams as data providers

### ISSUE THREAD

**Issue by aaronshaf** [marko-js/marko-async#10]
Would it be possible to use a stream as a data provider? Some possible uses:

- Use the streamed data in a marko iterator
- Or easier: Simply stream the text into an async-fragment

**Comment by aaronshaf** [comment]
Along these same lines, it'd be great to have Rx Observables as data providers.

**Comment by philidem** [comment]
@aaronshaf An async-fragment is associated with a stream so you should be able to pipe your stream to the async out steam. For example, something similar to:
```javascript
myStream.pipe(asyncOut);
```

I'll see if I can create a working example to demonstrate it's usage.

**Comment by patrick-steele-idem** [comment]
What @philidem suggested will work for a stream of HTML, but not a `Stream` in object mode. Also, if you are trying to achieve progressive HTML rendering then things won't quite work as expected since streamed HTML is buffered by default so that we can control chunking. When an async fragment is ended we do force a flush of the buffer but that means that the HTML won't be flushed until the incoming stream is ended. You can force a flush yourself, but that won't help if `client-reorder` is true since the async fragment must be completed before the DOM node can be moved into place.

It would be interesting to explore the following:

- Support a `Stream` as a data provider and have each `data` event be used as data for the nested content. In this case, we are assuming that the Stream is object mode and each chunk is a JavaScript object.
- After each `data` event: render the body of the async fragment using the event arg as the value for the local data variable.
- If `client-reorder` is enabled for the async fragment then render each chunk to a DOM node that gets inserted _before_ the placeholder (instead of replacing the placeholder). When the incoming stream is ended then the placeholder should be removed from the DOM.

This would really only be beneficial if your backend supports streaming data out. For example, instead waiting for entire JSON object of search results, the back end could, in theory, stream out groups of search results items as they come available and the front-end could flush the HTML for each group immediately after rendering.

@aaronshaf Were you interested in using a `Stream` in object mode?

**Comment by philidem** [comment]
@aaronshaf It's currently not possible to use streams via the `async-fragment` tag. If you're trying to render a non-Object Mode stream, the best approach for you might be to create a custom tag that uses a renderer similar to the following:
```javascript
module.exports = function renderer(input, out) {
    var asyncOut = out.beginAsync();
    input.stream.pipe(asyncOut);
};
```
In this example, `stream` would be an argument/attribute to the tag. For example:
`<render-stream stream="myStream" />`

---

If you're trying to render a stream of objects (Object Mode) then you would also create a custom tag as well. This approach might be a little more complicated so before I describe how to do that, please try to clarify your use case. Are you dealing with streams of characters/bytes or streams of objects?

**Comment by aaronshaf** [comment]
Thank you for the responses. My scenario:

1. Start rendering a marko template until the relevant fragment / tag

2. Stream in rows from a database

3. React.renderToString each row, and push it into the fragment

4. When #2 and #3 completed, render the rest of the marko template.

I think your example satisfies my scenario. How would asyncOut.end fit in the example? Thanks again!

**Comment by patrick-steele-idem** [comment]
@aaronshaf. That is pretty straightforward, but you would need to create a custom tag that controls the rendering of the rows using code similar to the following:

```javascript
var Row = require('./Row.jsx');

exports.renderer = function(input, out) {
    var rowsSource = geRowsRxObservable();

    var asyncOut = out.beginAsync();

    // For each row that comes back from the database (asynchronously):
    rowsSource.forEach(function(rowData) {
        // Use React to render the row:
        var rowHtml = React.renderToString(rowData);
        asyncOut.write(rowHtml);
    });

    // When there are no more rows end the async fragment
    source.done(function() {
        asyncOut.end();
    });
}
``` 

I only have basic knowledge of Rx, but hopefully that gives you the idea.

**Comment by patrick-steele-idem** [comment]
UPDATE: I fixed the usage of React in the above example

**Comment by aaronshaf** [comment]
@patrick-steele-idem, thanks again.

Are the contents sent to `asyncOut.write` are being streamed to the browser only after `asyncOut.end` is called?

**Comment by philidem** [comment]
@aaronshaf I'm not an expert on the underlying `asyncOut` stream, but I believe that output won't be sent to the browser until you call `asyncOut.end()` or `asyncOut.flush()` (`flush` is automatically called by `end`).

The `asyncOut` is designed to minimize the number of chunks that get sent to the browser because each chunk adds overhead. Typically, one async fragment will be one chunk. There is a trade-off between sending lots of little chunks and buffering output. If you want the data to be sent immediately then I would recommend calling `flush` frequently.

**Comment by unlight** [comment]
Hi, All.
I'm looking for feature as described @aaronshaf above: "Stream in rows from a database"
Example here, http://streaming-playground-unlight-1.c9.io/
Do not quite understand what the result of this conversion, need create custom tag or what?

**Comment by patrick-steele-idem** [comment]
@unlight Streaming rows from a database would be best done using a custom tag, but it could be done using a helper as long as the `out` variable is passed to the helper. Here's some sample code to illustrate how that could be done as a custom tag:

__src/components/app-database-rows/renderer.js__

```javascript
var rowTemplate = require('marko').load(require.resolve('./row.marko'));

exports.renderer = function(input, out) {
    var rowsStream = input.rowsStream;
    
    var asynOut = out.beginAsync();
    
    rowsStream
        .on('data', function(data) {
            // Render the <tr>...</tr> using a Marko template:
            rowTemplate.render(data, asyncOut);

            asyncOut.flush(); // Flush out the HTML. You may want to do
                              // this every n number of rows for better
                              // compression
        })
        .on('error', function(e) {
            asyncOut.error(e);
        })
        .on('end', function() {
            asyncOut.end();
        });
};
```

And then in your template the custom tag can be used as shown below:

```xml
<table>
    <app-database-rows rows-stream="${data.rowsStream}"/>
</table>
```

**Comment by patrick-steele-idem** [comment]
I'm doing some house keeping and closing old issues. If this is still an issue please feel free to reopen or open a new issue. Thanks!
